src/main/scala/com/databricks/spark/avro/AvroRelation.scala

```diff
@@ -40,8 +46,31 @@ case class AvroRelation(location: String)(@transient val sqlContext: SQLContext)
     convertedSchema
   }
 
+  def convertToSparkSQL(obj: Any): Any = {
+    obj match {
+      case u: Utf8 => u.toString
+      case m: Map[Any, Any] => m.map(x => (x._1.toString, convertToSparkSQL(x._2)))
+      case a: GenericData.Array[Any] => a.map(x => convertToSparkSQL(x))
```

Comment:
.map() is still doing the Scala lambda allocation in a per-element basis, which is something we should avoid. We should similarly avoid it within buildScan, inside the baseRdd lambda. 

We could leave this optimization for future work, though, as there are a couple other places we can improve if we go down that path. For instance, rather than using Row.fromSeq() for every row, we can construct a SpecificMutableRow which can be reused across rows. Additionally, we could use the schema to generate an array of functions which convert the expected Avro type to the Spark SQL type, rather than using runtime type checking.

These optimizations would likely be made in tandem with some sort of benchmark to ensure that they are in fact significant improvements, and to ensure future patches do not regress performance.